### PR TITLE
feat: Nine-slice

### PR DESCRIPTION
**Pull request by jyoung4242**
This doesn't have an issue number and doesn't close any issue numbers.

Scope: This is a new Graphic object

Files Impacted:

Created: \Excalibur\src\engine\Graphics\NineSlice.ts
Created: \Excalibur\src\spec\NineSliceSpec.ts
Created: \Excalibur\site\docs\04-graphics\04.6-nineslice.mdx

all existing tests passed locally

### CODE REVIEW

File: src/engine/Graphics/NineSlice.ts [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+        return Math.ceil(destinationSize / tilesize);
+      case NineSliceStretch.TileFit:
+        return Math.ceil(destinationSize / tilesize);
+    }
+  }
+
+  /**
+   * Returns the position and size of the tile
+   */
+  _calculateParams(
```
Comment (Joshuahelmle) [inline src/engine/Graphics/NineSlice.ts:522]: Should this method be part of  the public api?

File: src/engine/Graphics/NineSlice.ts [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+
+  /**
+   *  Draws 1 of the 9 tiles based on parameters passed in
+   *  context is the ExcaliburGraphicsContext from the _drawImage function
+   *  destinationSize is the size of the destination image as a vector (width,height)
+   *  targetCanvas is the canvas to draw to
+   *  hstrategy and vstrategy are the horizontal and vertical stretching strategies
+   *  marginW and marginH are optional margins for the 9 slice for positioning
+   */
+  _drawTile(
```
Comment (Joshuahelmle) [inline src/engine/Graphics/NineSlice.ts:161]: Should this be part of the public API?

File: src/spec/nineSliceSpec.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    ctx = new ex.ExcaliburGraphicsContext2DCanvas({ canvasElement });
+  });
+
+  it('can exist', () => {
+    // eslint-disable-next-line no-console
+    console.log('ex.NineSlice being ran');
+    expect(ex.NineSlice).toBeDefined();
+  });
+
+  it('generates a new id', () => {
```
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/spec/nineSliceSpec.ts:125]: good catch

File: src/spec/nineSliceSpec.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    jasmine.addAsyncMatchers(ExcaliburAsyncMatchers);
+    canvasElement = document.createElement('canvas');
+    canvasElement.width = 64;
+    canvasElement.height = 64;
+    ctx = new ex.ExcaliburGraphicsContext2DCanvas({ canvasElement });
+  });
+
+  it('can exist', () => {
+    // eslint-disable-next-line no-console
+    console.log('ex.NineSlice being ran');
```
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/spec/nineSliceSpec.ts:121]: nope, that's debugging, will fix

File: src/spec/nineSliceSpec.ts
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/spec/nineSliceSpec.ts:1]: shouldn't it?

File: src/engine/Graphics/NineSlice.ts (hunk 1 above)
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:522]: not based on my implementation

File: src/engine/Graphics/NineSlice.ts (hunk 2 above)
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:161]: nope, both _drawTile and _calculateParams are internal utility functions

File: src/spec/nineSliceSpec.ts
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/spec/nineSliceSpec.ts:1]: oh you mean the filename, yeah

File: src/engine/Graphics/NineSlice.ts (hunk 1 above)
Comment (Joshuahelmle) [reply src/engine/Graphics/NineSlice.ts:522]: in that case, the method should get a proper access modifier like private or protected, just prefixing it with underscore indicates that intention already but you're not enforcing it, without a modifier TS still leaves it as public

File: src/engine/Graphics/NineSlice.ts (hunk 2 above)
Comment (Joshuahelmle) [reply src/engine/Graphics/NineSlice.ts:161]: see other comment https://github.com/excaliburjs/Excalibur/pull/3241#discussion_r1797640168

File: src/engine/Graphics/NineSlice.ts (hunk 1 above)
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:522]: any reason that should impact functionality?

File: src/engine/Graphics/NineSlice.ts [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+  canvasD: HTMLCanvasElement;
+  canvasE: HTMLCanvasElement;
+  canvasF: HTMLCanvasElement;
+  canvasG: HTMLCanvasElement;
+  canvasH: HTMLCanvasElement;
+  canvasI: HTMLCanvasElement;
+
+  private _logger = Logger.getInstance();
+
+  constructor(public config: NineSliceConfig) {
```
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:70]: i made private

File: src/engine/Graphics/NineSlice.ts [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+  sourceSprite: HTMLImageElement;
+  canvasA: HTMLCanvasElement;
+  canvasB: HTMLCanvasElement;
+  canvasC: HTMLCanvasElement;
+  canvasD: HTMLCanvasElement;
+  canvasE: HTMLCanvasElement;
+  canvasF: HTMLCanvasElement;
+  canvasG: HTMLCanvasElement;
+  canvasH: HTMLCanvasElement;
+  canvasI: HTMLCanvasElement;
```
[the remark this answers is not in the document]
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:66]: fixed

File: src/engine/Graphics/NineSlice.ts (hunk 2 above)
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:161]: fixed

File: src/engine/Graphics/NineSlice.ts (hunk 1 above)
Comment (jyoung4242) [reply src/engine/Graphics/NineSlice.ts:522]: fixed

**Pull request merged**